test/token/ERC20/extensions/ERC4626.t.sol

```diff
@@ -1,18 +1,42 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.0;
 
-import "erc4626-tests/ERC4626.test.sol";
+import {ERC4626Test} from "erc4626-tests/ERC4626.test.sol";
 
-import {SafeCast} from "../../../../contracts/utils/math/SafeCast.sol";
-import {ERC20Mock} from "../../../../contracts/mocks/ERC20Mock.sol";
-import {ERC4626Mock} from "../../../../contracts/mocks/ERC4626Mock.sol";
+import {SafeCast} from "openzeppelin/utils/math/SafeCast.sol";
+import {ERC20} from "openzeppelin/token/ERC20/ERC20.sol";
+import {ERC4626} from "openzeppelin/token/ERC20/extensions/ERC4626.sol";
+
+import {ERC20Mock} from "openzeppelin/mocks/ERC20Mock.sol";
+import {ERC4626Mock} from "openzeppelin/mocks/ERC4626Mock.sol";
+import {ERC4626OffsetMock} from "openzeppelin/mocks/token/ERC4626OffsetMock.sol";
+
+contract ERC4626VaultOffsetMock is ERC4626OffsetMock {
+    constructor(
+        ERC20 underlying_,
+        uint8 offset_
+    ) ERC20("My Token Vault", "MTKNV") ERC4626(underlying_) ERC4626OffsetMock(offset_) {}
+}
 
 contract ERC4626StdTest is ERC4626Test {
+    ERC20 private _underlying = new ERC20Mock();
+
     function setUp() public override {
-        _underlying_ = address(new ERC20Mock());
+        _underlying_ = address(_underlying);
```

Comment:
I think this has to be `new ERC20Mock`. Otherwise, it makes further tests dependant on each other.
Not a problem right now.

Reply:
I see, so your approach is favoring execution time, right? (Deploy a mock ERC20 once and not on each fuzz run)

My worry comes from the following scenario:

1. We add a test for the `mint` function that requires approval of the underlying asset
2. Later, if we need the `mint` for another test, we might be adding more approval than what's needed (eg. an old test left some approval)

Might not be a big issue, but I'm not clear if the tradeoff between this and test performance is worth it.

Reply:
We have a limited understanding of how foundry testing works. If there are multiple `testXxxx` functions in a `XxxTest` contract, which of the following is true:

- A new (fresh) `XxxTest` is deployed for running each test
- A single `XxxTest` is deployed, but snapshot are used to ensure the state is "clean" at the begining of each `testXxx` function
- A single `XxxTest` contract is deployed, and some function (setUp?) is used to clean the contract between tests?

It looks like its not option 3.

Reply:
With option 1 or 2, what happens is:

- the `_underlyin = new ERC20Mock()` is executed as part of the implicit constructor.
- the `setUp()` function is then called, and initializes the `ERC4646Test` variable.
- we are ready to test

---

It feel like we could even do
```solidity
    ERC4626 private _vault = newERC4626Mock(new ERC20Mock());
    
    function setUp() public override {
        _underlying_ = _vault.asset();
        _vault_ = address(_vault);
        _delta_ = 0;
        _vaultMayBeEmpty = true;
        _unlimitedAmount = true;
    }
```
Would that work just as fine ?

Reply:
They look the same to me

![Captura de pantalla 2023-03-30 a la(s) 12 50 36](https://user-images.githubusercontent.com/33379285/228935273-9ee15578-f9d2-4ca5-92d5-73012da6688a.png)

These are the sources:

```solidity
contract CounterTest is Test {
    Counter public counter;

    function setUp() public {
        counter = new Counter();
        counter.setNumber(0);
    }

    function testSetNumber(uint256 x) public {
        counter.setNumber(x);
        assertEq(counter.number(), x);
    }
}
```

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "forge-std/Test.sol";
import "../src/Counter.sol";

contract CounterTest is Test {
    Counter public counter = new Counter();

    function setUp() public {
        counter.setNumber(0);
    }

    function testSetNumber(uint256 x) public {
        counter.setNumber(x);
        assertEq(counter.number(), x);
    }
}
```

Reply:
They look the same to me

![Captura de pantalla 2023-03-30 a la(s) 12 50 36](https://user-images.githubusercontent.com/33379285/228935273-9ee15578-f9d2-4ca5-92d5-73012da6688a.png)

These are the sources:

```solidity
contract CounterTest is Test {
    Counter public counter;

    function setUp() public {
        counter = new Counter();
        counter.setNumber(0);
    }

    function testSetNumber(uint256 x) public {
        counter.setNumber(x);
        assertEq(counter.number(), x);
    }
}
```

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "forge-std/Test.sol";
import "../src/Counter.sol";

contract CounterTest is Test {
    Counter public counter = new Counter();

    function setUp() public {
        counter.setNumber(0);
    }

    function testSetNumber(uint256 x) public {
        counter.setNumber(x);
        assertEq(counter.number(), x);
    }
}
```

Reply:
Well, `setUp` is an optional function invoked before **each** test case (i.e. a function prepended with `test`) is run. This is not something specific to Foundry but has its roots in hevm. You can see this [here](https://github.com/dapphub/dapptools/blob/master/src/hevm/src/EVM/UnitTest.hs#L125). So what happens in our case is the following:
1. The state of `ERC20 private _underlying = new ERC20Mock();` and any modification thereof via the _implicit_ constructor (i.e. part of the contract-level state variable definitions) remain the same across test functions.
2. Any modification of `_underlying` or any other variable within the `setUp` call is invoked before each test function. This means if you for instance deploy within the `setUp` function an `Ownable` contract, for which the `msg.sender` equals the `owner` by design, and you have a unit test that transfers the ownership multiple times, the state is reset if you call another test function, i.e. the `msg.sender` will be again the `owner` in the new test function at the beginning. All unit tests are one transaction and start with the post-`setUp` call state. Btw, that's one of the reasons why you can't test properly `selfdestruct` since the code is deleted at the end of a transaction. 

The reason why I didn't choose to use ` ERC4626 private _vault = new ERC4626Mock(new ERC20Mock());` directly, is that maybe we want to re-use `new ERC20Mock()` at some point.

Reply:
`new ERC20Mock()` would be accessible via `_vault.asset()`

Reply:
But the burning question is:

If we do what you propose AND if one of the tests mints some underlying tokens, will the following tests also have the token minted? We want all test to start with a "blank" state, and that includes having a brand new underlying token with no balances (in particular, no balance for the vault)

Reply:
> new ERC20Mock() would be accessible via _vault.asset()

But this returns an `address` type directly and not the contract object itself. Maybe we want to do something like that once:

```solidity
_underlying.mint(holder, type(uint16).max);
_underlying.approve(spender, type(uint256).max);
```

> But the burning question is:
> 
> If we do what you propose AND if one of the tests mints some underlying tokens, will the following tests also have the token minted? We want all test to start with a "blank" state, and that includes having a brand new underlying token with no balances (in particular, no balance for the vault)

The short answer is yes, it will have again the clean state (i.e. the post-`setUp` call state). I will make an example taken from my own unit tests [here](https://github.com/pcaversaccio/snekmate/blob/main/test/tokens/ERC721.t.sol) to illustrate this.

Take for instance [this](https://github.com/pcaversaccio/snekmate/blob/main/test/tokens/ERC721.t.sol#LL595-L604C6) unit test:

```solidity
function testBalanceOfCase1() public {
    address owner = makeAddr("owner");
    string memory uri1 = "my_awesome_nft_uri_1";
    string memory uri2 = "my_awesome_nft_uri_2";
    vm.startPrank(deployer);
    ERC721Extended.safe_mint(owner, uri1);
    ERC721Extended.safe_mint(owner, uri2);
    assertEq(ERC721Extended.balanceOf(owner), 2);
    vm.stopPrank();
}
```

Now further down, take the [following](https://github.com/pcaversaccio/snekmate/blob/main/test/tokens/ERC721.t.sol#LL1422-L1444C24) unit test which does another couple of NFT mints for instance.

```solidity
function testSafeMintSuccess() public {
        address owner = makeAddr("owner");
        string memory uri1 = "my_awesome_nft_uri_1";
        string memory uri2 = "my_awesome_nft_uri_2";
        string memory uri3 = "my_awesome_nft_uri_3";
        uint256 tokenId = 0;
        vm.startPrank(deployer);
        vm.expectEmit(true, true, true, false);
        emit Transfer(zeroAddress, owner, tokenId);
        vm.expectEmit(false, false, false, true);
        emit MetadataUpdate(tokenId);
        ERC721Extended.safe_mint(owner, uri1);
        vm.expectEmit(true, true, true, false);
        emit Transfer(zeroAddress, owner, tokenId + 1);
        vm.expectEmit(false, false, false, true);
        emit MetadataUpdate(tokenId + 1);
        ERC721Extended.safe_mint(owner, uri2);
        vm.expectEmit(true, true, true, false);
        emit Transfer(zeroAddress, owner, tokenId + 2);
        vm.expectEmit(false, false, false, true);
        emit MetadataUpdate(tokenId + 2);
        ERC721Extended.safe_mint(owner, uri3);
        vm.stopPrank();

        vm.startPrank(owner);
        vm.expectEmit(true, true, true, false);
        emit Transfer(owner, zeroAddress, tokenId + 2);
        ERC721Extended.burn(tokenId + 2);
        vm.stopPrank();

        vm.startPrank(deployer);
        vm.expectEmit(true, true, true, false);
        emit Transfer(zeroAddress, owner, tokenId + 3);
        vm.expectEmit(false, false, false, true);
        emit MetadataUpdate(tokenId + 3);
        ERC721Extended.safe_mint(owner, "");
        vm.stopPrank();
        assertEq(ERC721Extended.balanceOf(owner), 3);
        assertEq(ERC721Extended.totalSupply(), 3);
        assertEq(ERC721Extended.ownerOf(tokenId), owner);
        assertEq(ERC721Extended.ownerOf(tokenId + 1), owner);
        assertEq(ERC721Extended.ownerOf(tokenId + 3), owner);
    }
```

The above test would fail if it had the post `testBalanceOfCase1` state (since the NFTs would already exist). What I usually do is a `testInitalSetup` unit test that checks the post-`setUp` call. For instance, for my `ERC4626` (Vyper) implementation see [here](https://github.com/pcaversaccio/snekmate/blob/main/test/extensions/ERC4626.t.sol#LL182-L311C6). That's also where I e.g. test for the overflow problem. Does this clarify your question?

Reply:
That is also what I observed. Its just that this part is essential and is often overlooked. This is what @ernestognw meant when he said "makes further tests dependant on each other".

I guess snapshot are involved here.

Reply:
Fyi, if you want to handle snapshots directly yourself, you can use the `snapshot` cheatcode as well: https://book.getfoundry.sh/cheatcodes/snapshots. Do you feel overall uncomfortable with this change?